### ISSUE TITLE
Toolchain dependent deserialization

### ISSUE THREAD

**Issue by vlopes11** [rkyv/rkyv#416]
Dear colleagues; first and foremost, thanks for the amazing project!

I faced an unusual issue with rkyv. I serialized a data structure using `nightly-2023-05-22`, and attempted to deserialize it using a binary compiled on a different host that used a binary of a different toolchain: `nightly-2023-01-05`

I got the following error:

```
thread 'main' panicked at 'invalid notes: CheckBytesError(CheckBytesError(ValueCheckBytesError(CheckBytes { index: 0, error: StructCheckError { field_name: "note_type", inner: InvalidTag(254) } })))', src/main.rs:36:57
```

However, when I compile the deserialization binary with the same toolchain, it works just fine. I reduced the problem into a reproducible example

Cargo.toml
```toml
[package]
name = "rkyv-phoenix"
version = "0.1.0"
edition = "2021"

[dependencies]
phoenix-core = { version = "0.20.0-rc.0", features = ["rkyv-impl"] }
rkyv = "0.7"
```

src/main.rs
```rust
fn main() {
    let notes: Vec<u8> = vec![
        226, 47, 10, 53, 222, 231, 112, 244, 239, 44, 213, 213, 58, 60, 226, 43, 224, 230, 123, 2,
        41, 213, 92, 2, 173, 250, 140, 174, 243, 51, 199, 16, 157, 176, 185, 154, 205, 237, 68, 15,
        232, 200, 239, 218, 151, 41, 56, 69, 69, 241, 50, 176, 211, 96, 98, 180, 10, 192, 108, 100,
        6, 197, 192, 100, 95, 109, 209, 150, 81, 212, 131, 43, 154, 197, 11, 69, 139, 114, 68, 168,
        250, 60, 255, 89, 19, 252, 244, 29, 26, 200, 26, 59, 164, 211, 21, 115, 151, 55, 109, 95,
        189, 99, 129, 91, 29, 9, 126, 55, 134, 47, 121, 177, 169, 157, 248, 255, 152, 4, 238, 213,
        139, 103, 179, 112, 107, 143, 126, 15, 26, 182, 111, 197, 79, 53, 12, 166, 249, 65, 33,
        226, 144, 35, 142, 88, 187, 23, 233, 129, 19, 145, 66, 110, 174, 198, 3, 87, 15, 175, 109,
        86, 19, 45, 6, 190, 171, 142, 119, 66, 89, 172, 185, 186, 21, 65, 114, 226, 137, 0, 244,
        65, 157, 60, 241, 210, 219, 89, 237, 160, 183, 121, 152, 70, 93, 149, 83, 104, 134, 71,
        129, 170, 235, 236, 149, 208, 109, 104, 40, 216, 75, 148, 149, 248, 242, 252, 195, 47, 77,
        97, 25, 173, 177, 166, 77, 10, 36, 125, 68, 32, 88, 134, 85, 119, 120, 69, 29, 123, 72,
        131, 181, 186, 222, 57, 95, 130, 204, 3, 17, 44, 43, 107, 50, 28, 156, 163, 50, 14, 218,
        82, 25, 77, 118, 121, 194, 174, 121, 161, 151, 140, 12, 183, 198, 215, 79, 66, 197, 199,
        112, 180, 158, 248, 133, 116, 118, 119, 204, 81, 251, 101, 78, 69, 248, 9, 190, 60, 196,
        192, 104, 18, 139, 229, 129, 116, 111, 64, 15, 12, 94, 225, 19, 122, 131, 230, 165, 208,
        204, 87, 154, 5, 244, 113, 209, 120, 43, 106, 35, 243, 42, 160, 245, 167, 135, 104, 249,
        240, 44, 129, 34, 100, 55, 200, 118, 213, 195, 207, 235, 201, 123, 118, 86, 55, 238, 32,
        183, 105, 38, 21, 18, 85, 122, 88, 198, 82, 152, 181, 103, 252, 240, 190, 218, 77, 227,
        241, 158, 116, 184, 113, 5, 128, 75, 107, 123, 117, 186, 21, 175, 245, 43, 217, 3, 31, 158,
        62, 124, 152, 133, 0, 109, 129, 215, 124, 96, 121, 188, 75, 235, 85, 157, 15, 194, 40, 183,
        191, 218, 199, 60, 71, 27, 144, 16, 72, 11, 57, 136, 119, 132, 218, 226, 23, 232, 49, 33,
        151, 67, 213, 56, 63, 48, 205, 64, 200, 149, 85, 17, 238, 131, 59, 173, 91, 15, 0, 193,
        212, 114, 20, 105, 221, 173, 66, 44, 91, 216, 253, 66, 25, 61, 162, 150, 244, 222, 9, 255,
        125, 177, 82, 139, 48, 212, 52, 12, 18, 117, 26, 244, 145, 109, 73, 66, 109, 124, 135, 113,
        12, 52, 233, 201, 1, 124, 10, 40, 29, 214, 183, 16, 70, 33, 212, 64, 95, 165, 117, 241, 42,
        0, 0, 0, 0, 0, 0, 0, 0, 254, 255, 255, 255, 1, 0, 0, 0, 2, 72, 3, 0, 250, 183, 132, 88,
        245, 79, 188, 236, 239, 79, 140, 153, 111, 5, 197, 172, 89, 177, 36, 24, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 144, 253, 255, 255, 1, 0, 0, 0,
    ];

    rkyv::from_bytes::<Vec<phoenix_core::Note>>(&notes).expect("invalid notes");
}
```

To reproduce the problem:
```sh
cargo +nightly-2023-01-05 run
```

To run the deserialization with the same toolchain used for serialization (i.e. no errors):
```sh
cargo +nightly-2023-05-22 run
```

**Comment by djkoloski** [comment]
This is an unfortunate growing pain from rkyv's earlier days. You can turn on the `strict` feature to ensure that your format doesn't change with Rust toolchain upgrades. This will be the default starting with the next major revision.

**Comment by praveenperera** [comment]
@djkoloski we are seeing the same error `Unable to deserialize operation: CheckBytesError(CheckBytesError(InvalidTuple ...`

When upgrading from rust 1.64 to 1.71. We aren't using the `strict` feature flag. Any ideas? Thanks

**Comment by praveenperera** [comment]
@djkoloski to confirm this the `strict` feature would need to be enabled during serialization right? Enabling it afterwards for deserialization would not be enough?

**Comment by djkoloski** [comment]
That's correct, it must be enabled during serialization. It should be sufficient to load and re-serialize everything under `strict`. I don't have a good method for doing so right now, but if you also support a serde codec then you can use that as a side-channel to do so.

**Comment by praveenperera** [comment]
@djkoloski can this happen between stable rust version as well? say between 1.60 -> 1.64 or 1.64 -> 1.73 ?

**Comment by djkoloski** [comment]
If `strict` is on then it cannot happen between stable rust versions. Otherwise, yes it may occur.

**Comment by praveenperera** [comment]
Thanks @djkoloski, do you know what else could cause it. I am trying to deserialize some items that were serialized from an app compiled from rust 1.64. I tried using the same base docker image on the new app but I'm still getting this error.

`Unable to deserialize state: CheckBytesError(CheckBytesError(StructCheckError { field_name: "library", inner: StructCheckError { field_name: "timelines", inner: HashIndexError(ContextError(ArchiveError(Overrun { ptr: 0x7f1210071574, size: 17179868992, range: 0x7f121005bbe0..0x7f12100715c8 }))) } }`

I even downgraded to the rkyv version at the time (0.7)

Thanks

**Comment by djkoloski** [comment]
These stability issues should be resolved in 0.8. `strict` has been renamed to `stable_layout` and is on by default. All primitive types are now wrapped to ensure exact size and alignment, and when #456 is done we'll catch any cases where we don't have `#[repr(C)]` on provided types.
